ext/git.c

```diff
@@ -0,0 +1,339 @@
+    struct _git_metadata *git_metadata = pemalloc(sizeof(struct _git_metadata), 1);
+    git_metadata->source = source;
+    git_metadata->property_commit = commit_sha;
+    git_metadata->property_repository = repository_url;
+    zend_hash_add_ptr(&DDTRACE_G(git_metadata), cwd, git_metadata);
+}
+
+void inject_git_metadata(zval *carrier, zend_string *cwd) {
+    git_source source = inject_from_env(carrier);
+    if (source == GIT_SOURCE_NONE) {
+        source = inject_from_global_tags(carrier);
+    }
+    if (source == GIT_SOURCE_NONE) {
+        source = inject_from_git_dir(carrier, cwd);
+    }
+
+    if (source != GIT_SOURCE_NONE) {
+        ddtrace_git_metadata *git_metadata_obj = (ddtrace_git_metadata *) Z_OBJ_P(carrier);
+        zend_string *property_commit = Z_STR(git_metadata_obj->property_commit);
+        zend_string *property_repository = Z_STR(git_metadata_obj->property_repository);
+        zend_string_addref(property_commit);
+        zend_string_addref(property_repository);
+        cache_git_metadata(cwd, property_commit, property_repository, source);
+    } else {
+        zend_hash_add_ptr(&DDTRACE_G(git_metadata), cwd, &empty_git_metadata);
+    }
+}
+
+void replace_git_metadata(struct _git_metadata *git_metadata, zend_string *commit_sha, zend_string *repository_url) {
+    zend_string_release(git_metadata->property_commit);
+    zend_string_release(git_metadata->property_repository);
+    git_metadata->property_commit = commit_sha;
+    git_metadata->property_repository = repository_url;
+}
+
+void refresh_git_metadata_if_needed(zend_string *cwd, struct _git_metadata *git_metadata) {
+    zend_string *git_dir = find_git_dir(ZSTR_VAL(cwd));
+    if (!git_dir) {
+        return; // Should we replace git_metadata's properties by (NULL, NULL)?
+    }
+
+    zend_string *commit_sha = get_commit_sha(ZSTR_VAL(git_dir));
+    if (commit_sha && zend_string_equals(git_metadata->property_commit, commit_sha)) {
+        zend_string_release(commit_sha);
+    } else {
+        zend_string *repository_url = get_repository_url(ZSTR_VAL(git_dir));
+        replace_git_metadata(git_metadata, commit_sha, repository_url);
+    }
+
+    zend_string_release(git_dir);
+}
+
+void use_cached_metadata(zval *carrier, struct _git_metadata *git_metadata) {
+    zend_string_addref(git_metadata->property_commit);
+    zend_string_addref(git_metadata->property_repository);
+    add_git_info(carrier, git_metadata->property_commit, git_metadata->property_repository);
+}
+
+void update_git_metadata(void) {
+    zend_string *cwd;
+    struct _git_metadata *git_metadata;
+    ZEND_HASH_FOREACH_STR_KEY_PTR(&DDTRACE_G(git_metadata), cwd, git_metadata) {
+        refresh_git_metadata_if_needed(cwd, git_metadata);
+    }
+    ZEND_HASH_FOREACH_END();
+
+    request_cache_updated = true;
+}
+
+void ddtrace_inject_git_metadata(zval *carrier) {
+    if (!request_cache_updated) {
```

Comment:
When is this value ever reset?

Reply:
Well, it was in `ddtrace_git_metadata_rinit`, but my fat finger must have acted again and removed this line.

Reply:
May I propose `DDTRACE_G(git_object)` being a `zend_object *` (which you release at request end if non-NULL)? If it's NULL, then we proceed as usual. If it isn't, you do a ZVAL_OBJ_COPY to carrier? That we we actually persist the git metadata for a whole request.

Reply:
And so you'd remove both `git_metadata` and `git_request_cache_updated` from ddtrace_globals? right?

Reply:
We still need `git_metadata` for _cross-request_ caching. Storing the `zend_object *` is only allowed for within a request.